absl/testing/absltest.py

```diff
@@ -1223,10 +1223,15 @@ def _sorted_list_difference(expected, actual):
 
 def _walk_structure_for_problems(a, b, aname, bname, problem_list):
   """The recursive comparison behind assertSameStructure."""
-  if type(a) != type(b) and not (  # pylint: disable=unidiomatic-typecheck
-      isinstance(a, six.integer_types) and isinstance(b, six.integer_types)):
+  # pylint: disable=unidiomatic-typecheck
+  if type(a) != type(b) and not (
+      isinstance(a, six.integer_types) and
+      isinstance(b, six.integer_types)) and not (
+          (type(a) == collections.defaultdict and type(b) == dict) or
```

Comment:
Instead of special casing default dict, just special case all mappings. That is, if both a and b are mappings, go through the "compare mappings" code path. If their types differ, then go down this "things don't match and there's a problem" code path.

so do something like `... and isinstance(a, Mapping) != isisntance(b, Mapping))`

Reply:
Also, please add a comment down below, in the "walk over mappings" code that a or b could be a default dict, so care must be taken that they are not modified during traversal. Please add a test for that, too.